### PR TITLE
Intellij IDEA code style config

### PR DESCRIPTION
**Pull request by hnOsmium0001**
Just an enhancement for people who use IntelliJ IDEA to avoid making everybody setting the config themselves.
Unfortunately I don't use Eclipse anymore and don't have it installed, so I can't provide its code style config. If somebody has one they can also pull request them.

### CODE REVIEW

**Comment by MajorTuvok** [comment]
To my knowledge everyone of us uses IntelliJ

File: .gitignore [hunk 1]
```diff
@@ -3,6 +3,7 @@
 # folders
 !/gradle
 !/src
+!/ide
```
Comment (MajorTuvok) [inline .gitignore]: please don't open up the ide folder as a whole, but only the CodeStyle file... People like me will forget that at some point and accidently commit other IntelliJ Files.

File: ide/idea/CodeStyle.xml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<code_scheme name="Project" version="173">
+  <JavaCodeStyleSettings>
+    <option name="PACKAGES_TO_USE_IMPORT_ON_DEMAND">
```
Comment (MajorTuvok) [inline ide/idea/CodeStyle.xml]: ?!? Not needed here

**Comment by MichaelHillcox** [comment]
I wouldn't suggest this. If there is a java linter like eslint for js then I'd suggest that over a static config.

**Comment by MajorTuvok** [comment]
Well IntelliJ already does lint checks... And allows for auto formatting, which is configured by this.

**Comment by MichaelHillcox** [comment]
Yeah, but that's just idea, there are a few others that this would miss out on meaning that although we'd all be good, others that use eclipse for say wouldn't be following the same linting as us. Although, I agree that we need something like this as half our pr reviews are style related. I feel like there is a better way than using a config for a single ide

**Comment by hnOsmium0001** [comment]
Creating a code style config for eclipse too.

**Comment by MichaelHillcox** [comment]
What's holding up this pr?

**Comment by MajorTuvok** [comment]
Me forgetting to put my own config out there to be included

**Comment by MajorTuvok** [comment]
Here's what IntelliJ adds for me as Project wide Settings in contrast to my defaults...
```
<component name="ProjectCodeStyleConfiguration">
  <code_scheme name="Project" version="173">
    <JavaCodeStyleSettings>
      <option name="JD_ALIGN_PARAM_COMMENTS" value="false" />
      <option name="JD_ALIGN_EXCEPTION_COMMENTS" value="false" />
      <option name="JD_ADD_BLANK_AFTER_DESCRIPTION" value="false" />
      <option name="JD_KEEP_EMPTY_PARAMETER" value="false" />
      <option name="JD_KEEP_EMPTY_EXCEPTION" value="false" />
      <option name="JD_KEEP_EMPTY_RETURN" value="false" />
    </JavaCodeStyleSettings>
    <codeStyleSettings language="JAVA">
      <option name="KEEP_BLANK_LINES_IN_DECLARATIONS" value="1" />
      <option name="KEEP_BLANK_LINES_IN_CODE" value="1" />
      <option name="KEEP_BLANK_LINES_BETWEEN_PACKAGE_DECLARATION_AND_HEADER" value="1" />
      <option name="ELSE_ON_NEW_LINE" value="true" />
      <option name="ALIGN_MULTILINE_PARAMETERS_IN_CALLS" value="true" />
      <option name="ALIGN_MULTILINE_METHOD_BRACKETS" value="true" />
      <option name="SPACE_AROUND_UNARY_OPERATOR" value="true" />
      <option name="METHOD_CALL_CHAIN_WRAP" value="2" />
      <option name="KEEP_SIMPLE_BLOCKS_IN_ONE_LINE" value="true" />
      <option name="KEEP_SIMPLE_METHODS_IN_ONE_LINE" value="true" />
      <option name="KEEP_SIMPLE_LAMBDAS_IN_ONE_LINE" value="true" />
      <option name="KEEP_SIMPLE_CLASSES_IN_ONE_LINE" value="true" />
      <option name="ENUM_CONSTANTS_WRAP" value="2" />
    </codeStyleSettings>
  </code_scheme>
</component>
```

**Comment by MajorTuvok** [comment]
(I have absolutely no idea, what this `KEEP_EMPTY_EXCEPTION` thing is about)

**Comment by MajorTuvok** [comment]
@hnOsmium0001 any update?

**Comment by MichaelHillcox** [comment]
I'll be closing and marking this PR as stale if there continues to be no updates made within the next week ( as of 11th March )

**Comment by hnOsmium0001** [comment]
So the plan is to merge the two IDEA configs?

**Comment by MajorTuvok** [comment]
I thought so, if you agree with these settings (they should match the contribution guidelines...)

There's one thing that annoys me a bit: These simple Methods.
Yes it says that in the Contrubtion guidelines, but everytime I then call delegate or generate overrides, IntelliJ creates a Block of the form:
`private int m () {return super.()}`
All in the same line...

**Comment by MajorTuvok** [comment]
Everything else should be good

**Comment by MichaelHillcox** [comment]
`Yes it says that in the Contrubtion guidelines for completly empty brackets, but everytime I then call delegate or generate overrides, IntelliJ creates a Block of the form:
private int m () {return super.m()}
All in the same line...` I'm confused on your issue here

**Comment by MajorTuvok** [comment]
What I'm referring to is IntelliJ's generate facility, which generates one line Methods because of this:
`public int m() {return super.m():}`
instead of the expected
```
public int m() {
    return super.m();
}
```
because IntelliJ thinks, that that is a _trivial Method_ and according to the setting should be a one liner.
There doesn't seem to be a setting for empty Methods...

**Comment by MichaelHillcox** [comment]
No that is just a sudo inline it's not actually inline. Intelij just makes it look like it is

**Comment by MajorTuvok** [comment]
No... You can have alook with any text editor of your choice at it... It is an actual inline

One can already see that in IntelliJ - sudo inlines are marked green, and when you click on them they expand.

**Comment by MichaelHillcox** [comment]
Weird, doesn't happen for me

**Comment by MajorTuvok** [comment]
Really?
Why does my laptop hate me!!! -_-

**Comment by hnOsmium0001** [comment]
I believe there's no way to make IntelliJ remove simple braces for statement bodies without plugins.

**Comment by hnOsmium0001** [comment]
Request for Squash & Merge because many unnecessary commits were created due to accidents.

**Pull request merged**